be/src/runtime/batch_write/txn_state_cache.cpp

```diff
@@ -0,0 +1,467 @@
+    auto cache = _get_txn_cache(txn_id);
+    ASSIGN_OR_RETURN(auto entry, _get_txn_entry(cache, txn_id, true));
+    DCHECK(entry != nullptr);
+    entry->value().push_state(status, reason);
+    cache->release(entry);
+    return Status::OK();
+}
+
+StatusOr<TxnState> TxnStateCache::get_state(int64_t txn_id) {
+    auto cache = _get_txn_cache(txn_id);
+    ASSIGN_OR_RETURN(auto entry, _get_txn_entry(cache, txn_id, false));
+    if (entry == nullptr) {
+        return Status::NotFound("Transaction state not found");
+    }
+    auto txn_state = entry->value().txn_state();
+    cache->release(entry);
+    return txn_state;
+}
+
+StatusOr<TxnStateSubscriberPtr> TxnStateCache::subscribe_state(int64_t txn_id, const std::string& subscriber_name,
+                                                               const std::string& db, const std::string& tbl,
+                                                               const AuthInfo& auth) {
+    auto cache = _get_txn_cache(txn_id);
+    ASSIGN_OR_RETURN(auto entry, _get_txn_entry(cache, txn_id, true));
+    DCHECK(entry != nullptr);
+    bool trigger_poll = false;
+    entry->value().subscribe(trigger_poll);
+    TRACE_BATCH_WRITE << "create subscriber, txn_id: " << txn_id << ", name: " << subscriber_name << ", db: " << db
+                      << ", tbl: " << tbl << ", trigger_poll: " << trigger_poll;
+    if (trigger_poll) {
+        _txn_state_poller->submit({txn_id, db, tbl, auth}, config::merge_commit_txn_state_poll_interval_ms);
+    }
+    return std::make_unique<TxnStateSubscriber>(cache, entry, subscriber_name);
+}
+
+void TxnStateCache::set_capacity(size_t new_capacity) {
+    std::unique_lock<bthreads::BThreadSharedMutex> lock;
+    if (_stopped) {
+        return;
+    }
+    const size_t capacity_per_shard = (new_capacity + (kNumShards - 1)) / kNumShards;
+    for (auto& _shard : _shards) {
+        _shard->set_capacity(capacity_per_shard);
+    }
+    _capacity = new_capacity;
+}
+
+void TxnStateCache::stop() {
+    {
+        std::unique_lock<bthreads::BThreadSharedMutex> lock;
+        if (_stopped) {
+            return;
+        }
+        _stopped = true;
+    }
+    for (auto& cache : _shards) {
+        auto entries = cache->get_all_entries();
+        for (auto entry : entries) {
+            entry->value().stop();
+            cache->release(entry);
+        }
+    }
+    if (_txn_state_poller) {
+        _txn_state_poller->stop();
+    }
+    _poll_state_token->shutdown();
+}
+
+int32_t TxnStateCache::size() {
+    int32_t size = 0;
+    for (auto& cache : _shards) {
+        size += cache->size();
+    }
+    return size;
+}
+
+StatusOr<TxnStateDynamicCacheEntry*> TxnStateCache::_get_txn_entry(TxnStateDynamicCache* cache, int64_t txn_id,
+                                                                   bool create_if_not_exist) {
+    // use lock to avoid creating new entry after stopped
+    std::shared_lock<bthreads::BThreadSharedMutex> lock;
```

Comment:
This lock is only used to protect `_stopped`?

Reply:
yes